### ISSUE TITLE
how many epochs it take to train cogact from scratch

### ISSUE THREAD

**Issue by LiHaoHN** [microsoft/CogACT#5]
Hello, thank you for your work and open-source code! I want to ask how many epochs it takes to train cogact from scratch, It is written in the code that oxe_magic_soup_plus_minus configures 100 epochs, but I think it is very time-consuming (more time-consuming than training openvla?), you mentioned in the paper that you trained cogact 135K iterations, and your data is 22.5 million frames, and batchsize 256, does this mean you trained 135k/(22.5 M/256)= 1.536 epoch？
In addition, how do you judge that the model has finished the training? There is no token accuracy (similar to openvla) or val loss shown, is it training loss? If yes, how large the final training loss should be? Thank you again :)

**Comment by arnoldland** [comment]
Thank you for your attention!
1. In our config, we set epochs=100, but we manually terminated the training well before reaching this length of training time. The model we actually released was trained for 135K steps as stated in our paper. Due to the sampling mechanism of the dataloader, different datasets do not completely go through one full pass at the same time. However, The number of steps we used resulted in the main datasets (such as fractal20220817_data) going through approximately 2 epochs.

2. Good question! In fact, determining whether the robot policy has reached the optimal training step is a relatively difficult problem. Since we use diffusion loss to model the action sequences, there isn't a metric like token accuracy in OpenVLA to help us determine if the training should finish. For the pretraining stage, we suggest you can terminate the training when the average (Time Weighted EMA) training loss is about 0.05. As for the finetuning stage, we observed that training for approximately 15 epochs already yielded satisfactory performance. We have not yet thoroughly tested models finetuned for a longer duration in the real-world robot.

**Comment by LiHaoHN** [comment]
Get it! Thank you for your reply!

**Comment by LiHaoHN** [comment]
> Thank you for your attention!
> 
> 1. In our config, we set epochs=100, but we manually terminated the training well before reaching this length of training time. The model we actually released was trained for 135K steps as stated in our paper. Due to the sampling mechanism of the dataloader, different datasets do not completely go through one full pass at the same time. However, The number of steps we used resulted in the main datasets (such as fractal20220817_data) going through approximately 2 epochs.
> 2. Good question! In fact, determining whether the robot policy has reached the optimal training step is a relatively difficult problem. Since we use diffusion loss to model the action sequences, there isn't a metric like token accuracy in OpenVLA to help us determine if the training should finish. For the pretraining stage, we suggest you can terminate the training when the average (Time Weighted EMA) training loss is about 0.05. As for the finetuning stage, we observed that training for approximately 15 epochs already yielded satisfactory performance. We have not yet thoroughly tested models finetuned for a longer duration in the real-world robot.

I have another question. Under oxe_magic_soup_plus_minus ratio, what is the total length of the dataset? Is it consistent with the number of frames, like the 22.5M frames written in the paper, or is it smaller than this number?

**Comment by arnoldland** [comment]
We have solved this issue through offline communication. :)
